File: HTCommander.Core/radio/RadioDevInfo.cs
/*
Copyright 2026 Ylian Saint-Hilaire
Licensed under the Apache License, Version 2.0 (the "License");
http://www.apache.org/licenses/LICENSE-2.0
*/

using System;

namespace HTCommander
{
    public class RadioDevInfo
    {
        public byte[] raw;
        public int vendor_id;
        public int product_id;
        public int hw_ver;
        public int soft_ver;
        public bool support_radio;
        public bool support_medium_power;
        public bool fixed_loc_speaker_vol;
        public bool not_support_soft_power_ctrl;
        public bool have_no_speaker;
        public bool have_hm_speaker;
        public int region_count;
        public bool support_noaa;
        public bool gmrs;
        public bool support_vfo;
        public bool support_dmr;
        public int channel_count;
        public int freq_range_count;

        public RadioDevInfo(byte[] msg)
        {
            if (msg == null || msg.Length < 15) throw new ArgumentException("RadioDevInfo message too short (need >= 15 bytes)");
            raw = msg;
            vendor_id = msg[5];
            product_id = Utils.GetShort(msg, 6);
            hw_ver = msg[8];
            soft_ver = Utils.GetShort(msg, 9);
            support_radio = ((msg[11] & 0x80) != 0);
            support_medium_power = ((msg[11] & 0x40) != 0);
            fixed_loc_speaker_vol = ((msg[11] & 0x20) != 0);
            not_support_soft_power_ctrl = ((msg[11] & 0x10) != 0);
            have_no_speaker = ((msg[11] & 0x08) != 0);
            have_hm_speaker = ((msg[11] & 0x04) != 0);
            region_count = ((msg[11] & 0x03) << 4) + ((msg[12] & 0xF0) >> 4);
            support_noaa = ((msg[12] & 0x08) != 0);
            gmrs = ((msg[12] & 0x04) != 0);
            support_vfo = ((msg[12] & 0x02) != 0);
            support_dmr = ((msg[12] & 0x01) != 0);
            channel_count = msg[13];
            freq_range_count = (msg[14] & 0xF0) >> 4;
            if (freq_range_count > 8) freq_range_count = 8; // Cap at reasonable maximum
        }
    }
}

File: htcommander_flutter/lib/radio/models/radio_dev_info.dart
import 'dart:typed_data';
import '../binary_utils.dart';

/// Radio device information parsed from GAIA GET_DEV_INFO response.
/// Port of HTCommander.Core/radio/RadioDevInfo.cs
class RadioDevInfo {
  final Uint8List raw;
  final int vendorId;
  final int productId;
  final int hwVer;
  final int softVer;
  final bool supportRadio;
  final bool supportMediumPower;
  final bool fixedLocSpeakerVol;
  final bool notSupportSoftPowerCtrl;
  final bool haveNoSpeaker;
  final bool haveHmSpeaker;
  final int regionCount;
  final bool supportNoaa;
  final bool gmrs;
  final bool supportVfo;
  final bool supportDmr;
  final int channelCount;
  final int freqRangeCount;

  RadioDevInfo(Uint8List msg)
      : raw = msg,
        vendorId = msg[5],
        productId = BinaryUtils.getShort(msg, 6),
        hwVer = msg[8],
        softVer = BinaryUtils.getShort(msg, 9),
        supportRadio = (msg[11] & 0x80) != 0,
        supportMediumPower = (msg[11] & 0x40) != 0,
        fixedLocSpeakerVol = (msg[11] & 0x20) != 0,
        notSupportSoftPowerCtrl = (msg[11] & 0x10) != 0,
        haveNoSpeaker = (msg[11] & 0x08) != 0,
        haveHmSpeaker = (msg[11] & 0x04) != 0,
        regionCount = ((msg[11] & 0x03) << 4) + ((msg[12] & 0xF0) >> 4),
        supportNoaa = (msg[12] & 0x08) != 0,
        gmrs = (msg[12] & 0x04) != 0,
        supportVfo = (msg[12] & 0x02) != 0,
        supportDmr = (msg[12] & 0x01) != 0,
        channelCount = msg[13],
        freqRangeCount = _clampFreqRange((msg[14] & 0xF0) >> 4) {
    if (msg.length < 15) {
      throw ArgumentError(
          'RadioDevInfo message too short (need >= 15 bytes)');
    }
  }

  static int _clampFreqRange(int v) => v > 8 ? 8 : v;
}
